nemo/collections/common/data/lhotse/nemo_adapters.py

```diff
@@ -271,13 +273,24 @@ def _validate(self) -> None:
     def shard_ids(self) -> List[int]:
         return sorted(self.shard_id_to_manifest.keys())
 
-    def __iter__(self) -> Generator[Cut, None, None]:
-        shard_ids = self.shard_ids
-
-        if self.shuffle_shards:
-            seed = resolve_seed(self.shard_seed)
-            random.Random(seed).shuffle(shard_ids)
-
+    def _iter_random_read(self, shard_ids) -> Generator[tuple[dict, bytes], None, None]:
+        for sid in shard_ids:
+            shard_manifest = self.shard_id_to_manifest[sid]
+            tar_path = self.shard_id_to_tar_path[sid]
+            with tarfile.open(fileobj=open_best(tar_path, mode="rb"), mode="r") as tar:
+                for data in shard_manifest:
+                    try:
+                        tar_info = tar.getmember(data["audio_filepath"])
```

Comment:
Is `getmember` forcing a linear scan of the tarfile for every audio file read? It might be nice to include in general in Nemo some helpers for tarred dataset, where a separate index file with byte start offset/file size can be pre-computed and used for direct seek/read without openning/scanning the tarfile

Reply:
Yes, I believe it does a linear search. 
The optimization idea sounds good; I'll think about the implementation.

Reply:
Here is a method of getting the offsets from the tar member (for populating the index file): https://stackoverflow.com/a/43054257/445810 it is undocumented in cPython, but known to work. If such an index file is available e.g. we can check if `ourfile.tar` has next to it a `ourfile.tar.json` - and then use it for loading the index. At least having an option for this would be great. And not only for LHotse, but as a general dataset class in Nemo. A more advanced option is allowing a collection of tars/indexes to be loaded. This is useful when we have a dataset cut in several parts/shards for simplifying downloading these shards from s3 (better to download several files of 2Gb than a gigantic single file of 100Gb - for the cases if network is not perfect).

Reply:
There is a `wids` library in `webdataset` that does this, but I don't have first-hand experience with that. Perhaps adding explicit support for this would be desirable... but I think that's easily a separate full PR.

Reply:
An alternative to random reading is to download the whole shard into CPU memory sequentially and immediately discard the files that are not in the manifest. Maybe we should have gone that way instead. Hard to say..